src/libraries/Common/src/Roslyn/SyntaxValueProvider_ForAttributeWithSimpleName.cs

```diff
@@ -0,0 +1,295 @@
+            /*.WithTrackingName("compilationUnitAndGlobalAliases_ForAttribute")*/;
+
+        // For each pair of compilation unit + global aliases, walk the compilation unit
+        var result = compilationUnitAndGlobalAliasesProvider
+            .SelectMany((globalAliasesAndCompilationUnit, cancellationToken) => GetMatchingNodes(
+                syntaxHelper, globalAliasesAndCompilationUnit.Right, globalAliasesAndCompilationUnit.Left, simpleName, predicate, cancellationToken))
+            /*.WithTrackingName("result_ForAttribute")*/;
+
+        return result;
+
+        static GlobalAliases getGlobalAliasesInCompilationUnit(
+            SyntaxNode compilationUnit)
+        {
+            Debug.Assert(compilationUnit is ICompilationUnitSyntax);
+            var globalAliases = new Aliases(Span<(string aliasName, string symbolName)>.Empty);
+
+            CSharpSyntaxHelper.Instance.AddAliases(compilationUnit, ref globalAliases, global: true);
+
+            return GlobalAliases.Create(globalAliases.AsSpan().ToImmutableArray());
+        }
+    }
+
+    private static ImmutableArray<SyntaxNode> GetMatchingNodes(
+        ISyntaxHelper syntaxHelper,
+        GlobalAliases globalAliases,
+        SyntaxNode compilationUnit,
+        string name,
+        Func<SyntaxNode, CancellationToken, bool> predicate,
+        CancellationToken cancellationToken)
+    {
+        Debug.Assert(compilationUnit is ICompilationUnitSyntax);
+
+        var isCaseSensitive = syntaxHelper.IsCaseSensitive;
+        var comparison = isCaseSensitive ? StringComparison.Ordinal : StringComparison.OrdinalIgnoreCase;
+
+        // As we walk down the compilation unit and nested namespaces, we may encounter additional using aliases local
+        // to this file. Keep track of them so we can determine if they would allow an attribute in code to bind to the
+        // attribute being searched for.
+        var localAliases = new Aliases(Span<(string, string)>.Empty);
+        var nameHasAttributeSuffix = name.HasAttributeSuffix(isCaseSensitive);
+
+        // Used to ensure that as we recurse through alias names to see if they could bind to attributeName that we
+        // don't get into cycles.
+        var seenNames = new ValueListBuilder<string>(Span<string>.Empty);
+        var results = new ValueListBuilder<SyntaxNode>(Span<SyntaxNode>.Empty);
+        var attributeTargets = new ValueListBuilder<SyntaxNode>(Span<SyntaxNode>.Empty);
+
+        recurse(compilationUnit, ref localAliases, ref seenNames, ref results, ref attributeTargets);
+
+        if (results.Length == 0)
+            return ImmutableArray<SyntaxNode>.Empty;
+
+        return results.AsSpan().ToArray().Distinct().ToImmutableArray();
+
+        void recurse(
+            SyntaxNode node,
+            ref Aliases localAliases,
+            ref ValueListBuilder<string> seenNames,
+            ref ValueListBuilder<SyntaxNode> results,
+            ref ValueListBuilder<SyntaxNode> attributeTargets)
```

Comment:
not being able to reference ref-struct-locals from local-funcs sucks.  @jaredpar  can we fix this in the language :D